packages/interactivity/src/utils.ts

```diff
@@ -374,3 +392,26 @@ export const isPlainObject = (
 			typeof candidate === 'object' &&
 			candidate.constructor === Object
 	);
+
+/**
+ * Indicates that the passed `callback` requires synchronous access to the event object.
+ *
+ * @param callback The event callback.
+ * @return Wrapped event callback.
+ */
+export function withSyncEvent( callback: Function ): SyncAwareFunction {
+	let wrapped: SyncAwareFunction;
+
+	if ( callback?.constructor?.name === 'GeneratorFunction' ) {
+		wrapped = function* ( this: any, ...args: any[] ) {
+			yield* callback.apply( this, args );
+		};
+	} else {
+		wrapped = function ( this: any, ...args: any[] ) {
+			return callback.apply( this, args );
+		};
+	}
+
+	wrapped.sync = true;
+	return wrapped;
+}
```

Comment:
Why don't you simply add the `sync` property to the original function? Do you want to prevent modifying it?

```suggestion
export function withSyncEvent( callback: Function ): SyncAwareFunction {
	( callback as SyncAwareFunction ).sync = true;
	return callback as SyncAwareFunction;
}
```

Reply:
Great idea! I thought initially that wrapping it was safer than directly altering it, but since the `callback` function would be usually just declared anonymously and immediately passed to `withSyncEvent`, there's probably no side effects of directly altering it.

Most importantly, I agree with your suggestion because it makes the logic much simpler and more error-proof, maintaining whichever type the original function had before passing it to `withSyncEvent`, instead of trying to recreate/guess it.

I've amended this in https://github.com/WordPress/gutenberg/pull/68097/commits/f0fa92afaaca9b625a341610564c73f0fb9ad2a3, alongside minor wording adjustments based on the change.

Reply:
> Great idea! I thought initially that wrapping it was safer than directly altering it, but since the `callback` function would be usually just declared anonymously and immediately passed to `withSyncEvent`, there's probably no side effects of directly altering it.

The key word here is _usually_, right? You could define a named function in the module and then reference it here in `withSyncEvent()`. What if you then use that same function for an async event? Then it would eventually result in that code not being able run asynchronously, since the function has the `sync` flag set on it.

Reply:
Yeah I agree it wouldn't be 100% guaranteed. But given the wrapping requires other quirks that also don't make it 100% reliable (not every function is a plain function or has constructor `GeneratorFunction`), I think avoiding the wrapping is safer for the vast majority of use-cases.

Specifically to your example, I have a hard time seeing that being common: First, because if the function is the same and needs synchronous `event` access, why would it elsewhere _not_ access a synchronous `event` property? Second, all the Interactivity API examples and Core code define the actions right on the object, so it's not encouraged to do it differently anywhere.

Reply:
I also think that the likelihood of the same function being executed synchronously and asynchronously at the same time must be really low, because it would have to include conditional code and depend on the place of the DOM where it is executed. It is not impossible and I'm sure there is some case, but in that case, maybe it would even be safer to always run it synchronously.

That said, I don't oppose leaving that part of the code as it is. Up to you 🙂